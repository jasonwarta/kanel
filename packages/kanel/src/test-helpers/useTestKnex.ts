import knex, { type Knex } from "knex";
import { afterAll, beforeAll } from "vitest";

import usePostgresContainer from "./usePostgresContainer";

type Connection = {
  host: string;
  port: number;
  password: string;
  user: string;
  database: string;
};

const useTestKnex = (): readonly [
  () => Knex<any, any[]>,
  string,
  () => Connection,
] => {
  let knexInstance: Knex;
  const databaseName = `test_${Math.ceil(Math.random() * 1000)}`;
  let connection: Connection;

  const getContainer = usePostgresContainer();

  beforeAll(async () => {
    const container = getContainer();
    connection = {
      host: container.getHost(),
      port: container.getMappedPort(5432),
      password: "postgres",
      user: "postgres",
      database: "postgres",
    };

    const setupKnexInstance = knex({
      client: "postgres",
      connection,
    });
    await setupKnexInstance.raw("create database ??", [databaseName]);
    await setupKnexInstance.destroy();

    connection.database = databaseName;
    knexInstance = knex({
      client: "postgres",
      connection,
    });
  });

  afterAll(async () => {
    const container = getContainer();
    const connection = {
      host: container.getHost(),
      port: container.getMappedPort(5432),
      password: "postgres",
      user: "postgres",
    };

    const setupKnexInstance = knex({
      client: "postgres",
      connection: { ...connection, database: "postgres" },
    });

    setupKnexInstance
      .raw(`drop database ${databaseName} with (force)`)
      .then(() => setupKnexInstance.destroy());

    await knexInstance.destroy();
  });

  return [() => knexInstance, databaseName, () => connection] as const;
};

export default useTestKnex;
