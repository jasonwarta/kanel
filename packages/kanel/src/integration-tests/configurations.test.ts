import { describe, expect, it, vi } from "vitest";
import { recase } from "@kristiandupont/recase";
import { tryParse } from "tagged-comment-parser";
import { join } from "path";

import type { Config } from "../config-types";
import processDatabase from "../processDatabase";
import useTestKnex from "../test-helpers/useTestKnex";
import useSchema from "../test-helpers/useSchema";

vi.mock("../writeFile", () => ({
  default: vi.fn(),
}));

import writeFile from "../writeFile";
const mockedWriteFile = vi.mocked(writeFile);

function getResults(): { [fullPath: string]: string[] } {
  return mockedWriteFile.mock.calls.reduce(
    (acc, [args]) => {
      acc[args.fullPath] = args.lines;
      return acc;
    },
    {} as Record<string, string[]>,
  );
}

const outputPath = "./example/models";
const toPascalCase = recase("snake", "pascal");

const config: Partial<Config> = {
  outputPath,
  resolveViews: true,
  preDeleteOutputFolder: false,

  // Add a comment about the entity that the type represents above each type.
  getMetadata: (details, generateFor) => {
    const { comment: strippedComment } = tryParse(details.comment);
    const isAgentNoun = ["initializer", "mutator"].includes(generateFor);

    const relationComment = isAgentNoun
      ? `Represents the ${generateFor} for the ${details.kind} ${details.schemaName}.${details.name}`
      : `Represents the ${details.kind} ${details.schemaName}.${details.name}`;

    const suffix = isAgentNoun ? `_${generateFor}` : "";

    return {
      name: toPascalCase(details.name + suffix),
      comment: [relationComment, ...(strippedComment ? [strippedComment] : [])],
      path: join(outputPath, toPascalCase(details.name)),
    };
  },

  // Add a comment that says what the type of the column/attribute is in the database.
  getPropertyMetadata: (property, _details, generateFor) => {
    const { comment: strippedComment } = tryParse(property.comment);

    return {
      name: property.name,
      comment: [
        `Database type: ${property.expandedType}`,
        ...(generateFor === "initializer" && property.defaultValue
          ? [`Default value: ${property.defaultValue}`]
          : []),
        ...(strippedComment ? [strippedComment] : []),
      ],
    };
  },

  getRoutineMetadata: (details, _instantiatedConfig) => ({
    parametersName: `${details.name}_params`,
    parameters: details.parameters.map(({ name }) => ({
      name,
      comment: [],
    })),
    returnTypeName: `${details.name}_return_type`,
    returnTypeComment: [`Return type for ${details.name}`],
    path: join(outputPath, details.name),
  }),

  // This implementation will generate flavored instead of branded types.
  // See: https://spin.atomicobject.com/2018/01/15/typescript-flexible-nominal-typing/
  generateIdentifierType: (c, d) => {
    // Id columns are already prefixed with the table name, so we don't need to add it here
    const name = toPascalCase(c.name);

    return {
      declarationType: "typeDeclaration",
      name,
      exportAs: "named",
      typeDefinition: [`number & { __flavor?: '${name}' }`],
      comment: [`Identifier type for ${d.name}`],
    };
  },

  customTypeMap: {
    // A text search vector could be stored as a set of strings. See Film.ts for an example.
    "pg_catalog.tsvector": "Set<string>",

    // The bytea package (https://www.npmjs.com/package/postgres-bytea) could be used for byte arrays.
    // See Staff.ts for an example.
    "pg_catalog.bytea": {
      name: "bytea",
      typeImports: [
        {
          name: "bytea",
          path: "postgres-bytea",
          isAbsolute: true,
          isDefault: true,
          importAsType: false,
        },
      ],
    },

    // If you want to use BigInt for bigserial columns, you can use the following.
    "pg_catalog.int8": "BigInt",

    // Columns with the following types would probably just be strings in TypeScript.
    "pg_catalog.bpchar": "string",
    "public.citext": "string",
  },
};

describe("Integration tests for various configurations", () => {
  const [getKnex, _, getConnection] = useTestKnex();
  useSchema(getKnex, "test");

  it("should respect the configuration", async () => {
    const db = getKnex();
    const sql = `
      -- Enable extensions
      create extension if not exists citext;
      
      -- Basic tables with relationships
      create table test.some_table (id integer);
      create table test.some_other_table (id integer primary key);
      alter table test.some_table add column some_other_table_id integer references test.some_other_table (id);
      
      -- Custom types
      create type test.some_enum as enum ('value1', 'value2');
      create type test.some_range as range (subtype = test.some_enum);
      create domain test.some_domain as text;
      create type test.some_composite as (name text, value integer);
      
      -- Table with custom type columns (testing customTypeMap)
      create table test.custom_types_table (
        id serial primary key,
        search_vector tsvector,
        binary_data bytea,
        big_number bigint,
        fixed_char char(10),
        case_insensitive_text citext,
        enum_column test.some_enum,
        domain_column test.some_domain,
        range_column test.some_range,
        composite_column test.some_composite,
        text_array text[],
        enum_array test.some_enum[],
        int_array integer[],
        created_at timestamp with time zone default now(),
        updated_at timestamp with time zone default current_timestamp,
        is_active boolean default true,
        priority integer default 0
      );
      
      -- Table with complex relationships
      create table test.complex_table (
        id serial primary key,
        custom_types_id integer references test.custom_types_table (id),
        some_other_table_id integer references test.some_other_table (id),
        name varchar(255) not null,
        description text,
        metadata jsonb,
        tags text[] default '{}',
        created_at timestamp with time zone default now()
      );
      
      -- Views
      create view test.some_view as select * from test.some_table;
      create materialized view test.some_materialized_view as select * from test.some_table;
      
      -- Functions and procedures
      create function test.some_function() returns void language sql as $$ $$;
      create procedure test.some_procedure() language sql as $$ $$;
    `;
    await db.raw(sql);

    await processDatabase({
      connection: getConnection(),
      ...config,
    });

    expect(getResults()).toMatchSnapshot();
  });
});
