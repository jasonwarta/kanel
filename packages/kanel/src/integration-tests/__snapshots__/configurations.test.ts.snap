// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Integration tests for various configurations > should respect the configuration 1`] = `
{
  "example/models/ComplexTable.ts": [
    "// @generated",
    "// This file is automatically generated by Kanel. Do not modify manually.",
    "",
    "import type { Id } from './CustomTypesTable';",
    "import type { Id } from './SomeOtherTable';",
    "",
    "/** Identifier type for complex_table */",
    "export type Id = number & { __flavor?: 'Id' };",
    "",
    "/** Represents the table test.complex_table */",
    "export default interface ComplexTable {",
    "  /** Database type: pg_catalog.int4 */",
    "  id: Id;",
    "",
    "  /** Database type: pg_catalog.int4 */",
    "  custom_types_id: Id | null;",
    "",
    "  /** Database type: pg_catalog.int4 */",
    "  some_other_table_id: Id | null;",
    "",
    "  /** Database type: pg_catalog.varchar */",
    "  name: string;",
    "",
    "  /** Database type: pg_catalog.text */",
    "  description: string | null;",
    "",
    "  /** Database type: pg_catalog.jsonb */",
    "  metadata: unknown | null;",
    "",
    "  /** Database type: pg_catalog.text[] */",
    "  tags: string[] | null;",
    "",
    "  /** Database type: pg_catalog.timestamptz */",
    "  created_at: Date | null;",
    "}",
    "",
    "/** Represents the initializer for the table test.complex_table */",
    "export interface ComplexTableInitializer {",
    "  /**",
    "   * Database type: pg_catalog.int4",
    "   * Default value: nextval('test.complex_table_id_seq'::regclass)",
    "   */",
    "  id?: Id;",
    "",
    "  /** Database type: pg_catalog.int4 */",
    "  custom_types_id?: Id | null;",
    "",
    "  /** Database type: pg_catalog.int4 */",
    "  some_other_table_id?: Id | null;",
    "",
    "  /** Database type: pg_catalog.varchar */",
    "  name: string;",
    "",
    "  /** Database type: pg_catalog.text */",
    "  description?: string | null;",
    "",
    "  /** Database type: pg_catalog.jsonb */",
    "  metadata?: unknown | null;",
    "",
    "  /**",
    "   * Database type: pg_catalog.text[]",
    "   * Default value: '{}'::text[]",
    "   */",
    "  tags?: string[] | null;",
    "",
    "  /**",
    "   * Database type: pg_catalog.timestamptz",
    "   * Default value: now()",
    "   */",
    "  created_at?: Date | null;",
    "}",
    "",
    "/** Represents the mutator for the table test.complex_table */",
    "export interface ComplexTableMutator {",
    "  /** Database type: pg_catalog.int4 */",
    "  id?: Id;",
    "",
    "  /** Database type: pg_catalog.int4 */",
    "  custom_types_id?: Id | null;",
    "",
    "  /** Database type: pg_catalog.int4 */",
    "  some_other_table_id?: Id | null;",
    "",
    "  /** Database type: pg_catalog.varchar */",
    "  name?: string;",
    "",
    "  /** Database type: pg_catalog.text */",
    "  description?: string | null;",
    "",
    "  /** Database type: pg_catalog.jsonb */",
    "  metadata?: unknown | null;",
    "",
    "  /** Database type: pg_catalog.text[] */",
    "  tags?: string[] | null;",
    "",
    "  /** Database type: pg_catalog.timestamptz */",
    "  created_at?: Date | null;",
    "}",
    "",
  ],
  "example/models/CustomTypesTable.ts": [
    "// @generated",
    "// This file is automatically generated by Kanel. Do not modify manually.",
    "",
    "import bytea from 'postgres-bytea';",
    "import type { default as SomeEnum } from './SomeEnum';",
    "import type { default as SomeDomain } from './SomeDomain';",
    "import type { default as SomeRange } from './SomeRange';",
    "import type { default as SomeComposite } from './SomeComposite';",
    "",
    "/** Identifier type for custom_types_table */",
    "export type Id = number & { __flavor?: 'Id' };",
    "",
    "/** Represents the table test.custom_types_table */",
    "export default interface CustomTypesTable {",
    "  /** Database type: pg_catalog.int4 */",
    "  id: Id;",
    "",
    "  /** Database type: pg_catalog.tsvector */",
    "  search_vector: Set<string> | null;",
    "",
    "  /** Database type: pg_catalog.bytea */",
    "  binary_data: bytea | null;",
    "",
    "  /** Database type: pg_catalog.int8 */",
    "  big_number: BigInt | null;",
    "",
    "  /** Database type: pg_catalog.bpchar */",
    "  fixed_char: string | null;",
    "",
    "  /** Database type: public.citext */",
    "  case_insensitive_text: string | null;",
    "",
    "  /** Database type: test.some_enum */",
    "  enum_column: SomeEnum | null;",
    "",
    "  /** Database type: test.some_domain */",
    "  domain_column: SomeDomain | null;",
    "",
    "  /** Database type: test.some_range */",
    "  range_column: SomeRange | null;",
    "",
    "  /** Database type: test.some_composite */",
    "  composite_column: SomeComposite | null;",
    "",
    "  /** Database type: pg_catalog.text[] */",
    "  text_array: string[] | null;",
    "",
    "  /** Database type: test.some_enum[] */",
    "  enum_array: SomeEnum[] | null;",
    "",
    "  /** Database type: pg_catalog.int4[] */",
    "  int_array: number[] | null;",
    "",
    "  /** Database type: pg_catalog.timestamptz */",
    "  created_at: Date | null;",
    "",
    "  /** Database type: pg_catalog.timestamptz */",
    "  updated_at: Date | null;",
    "",
    "  /** Database type: pg_catalog.bool */",
    "  is_active: boolean | null;",
    "",
    "  /** Database type: pg_catalog.int4 */",
    "  priority: number | null;",
    "}",
    "",
    "/** Represents the initializer for the table test.custom_types_table */",
    "export interface CustomTypesTableInitializer {",
    "  /**",
    "   * Database type: pg_catalog.int4",
    "   * Default value: nextval('test.custom_types_table_id_seq'::regclass)",
    "   */",
    "  id?: Id;",
    "",
    "  /** Database type: pg_catalog.tsvector */",
    "  search_vector?: Set<string> | null;",
    "",
    "  /** Database type: pg_catalog.bytea */",
    "  binary_data?: bytea | null;",
    "",
    "  /** Database type: pg_catalog.int8 */",
    "  big_number?: BigInt | null;",
    "",
    "  /** Database type: pg_catalog.bpchar */",
    "  fixed_char?: string | null;",
    "",
    "  /** Database type: public.citext */",
    "  case_insensitive_text?: string | null;",
    "",
    "  /** Database type: test.some_enum */",
    "  enum_column?: SomeEnum | null;",
    "",
    "  /** Database type: test.some_domain */",
    "  domain_column?: SomeDomain | null;",
    "",
    "  /** Database type: test.some_range */",
    "  range_column?: SomeRange | null;",
    "",
    "  /** Database type: test.some_composite */",
    "  composite_column?: SomeComposite | null;",
    "",
    "  /** Database type: pg_catalog.text[] */",
    "  text_array?: string[] | null;",
    "",
    "  /** Database type: test.some_enum[] */",
    "  enum_array?: SomeEnum[] | null;",
    "",
    "  /** Database type: pg_catalog.int4[] */",
    "  int_array?: number[] | null;",
    "",
    "  /**",
    "   * Database type: pg_catalog.timestamptz",
    "   * Default value: now()",
    "   */",
    "  created_at?: Date | null;",
    "",
    "  /**",
    "   * Database type: pg_catalog.timestamptz",
    "   * Default value: CURRENT_TIMESTAMP",
    "   */",
    "  updated_at?: Date | null;",
    "",
    "  /**",
    "   * Database type: pg_catalog.bool",
    "   * Default value: true",
    "   */",
    "  is_active?: boolean | null;",
    "",
    "  /**",
    "   * Database type: pg_catalog.int4",
    "   * Default value: 0",
    "   */",
    "  priority?: number | null;",
    "}",
    "",
    "/** Represents the mutator for the table test.custom_types_table */",
    "export interface CustomTypesTableMutator {",
    "  /** Database type: pg_catalog.int4 */",
    "  id?: Id;",
    "",
    "  /** Database type: pg_catalog.tsvector */",
    "  search_vector?: Set<string> | null;",
    "",
    "  /** Database type: pg_catalog.bytea */",
    "  binary_data?: bytea | null;",
    "",
    "  /** Database type: pg_catalog.int8 */",
    "  big_number?: BigInt | null;",
    "",
    "  /** Database type: pg_catalog.bpchar */",
    "  fixed_char?: string | null;",
    "",
    "  /** Database type: public.citext */",
    "  case_insensitive_text?: string | null;",
    "",
    "  /** Database type: test.some_enum */",
    "  enum_column?: SomeEnum | null;",
    "",
    "  /** Database type: test.some_domain */",
    "  domain_column?: SomeDomain | null;",
    "",
    "  /** Database type: test.some_range */",
    "  range_column?: SomeRange | null;",
    "",
    "  /** Database type: test.some_composite */",
    "  composite_column?: SomeComposite | null;",
    "",
    "  /** Database type: pg_catalog.text[] */",
    "  text_array?: string[] | null;",
    "",
    "  /** Database type: test.some_enum[] */",
    "  enum_array?: SomeEnum[] | null;",
    "",
    "  /** Database type: pg_catalog.int4[] */",
    "  int_array?: number[] | null;",
    "",
    "  /** Database type: pg_catalog.timestamptz */",
    "  created_at?: Date | null;",
    "",
    "  /** Database type: pg_catalog.timestamptz */",
    "  updated_at?: Date | null;",
    "",
    "  /** Database type: pg_catalog.bool */",
    "  is_active?: boolean | null;",
    "",
    "  /** Database type: pg_catalog.int4 */",
    "  priority?: number | null;",
    "}",
    "",
  ],
  "example/models/SomeComposite.ts": [
    "// @generated",
    "// This file is automatically generated by Kanel. Do not modify manually.",
    "",
    "/** Represents the compositeType test.some_composite */",
    "export default interface SomeComposite {",
    "  /** Database type: pg_catalog.text */",
    "  name: string | null;",
    "",
    "  /** Database type: pg_catalog.int4 */",
    "  value: number | null;",
    "}",
    "",
  ],
  "example/models/SomeDomain.ts": [
    "// @generated",
    "// This file is automatically generated by Kanel. Do not modify manually.",
    "",
    "/** Represents the domain test.some_domain */",
    "type SomeDomain = string;",
    "",
    "export default SomeDomain;",
    "",
  ],
  "example/models/SomeEnum.ts": [
    "// @generated",
    "// This file is automatically generated by Kanel. Do not modify manually.",
    "",
    "/** Represents the enum test.some_enum */",
    "enum SomeEnum {",
    "  value1 = 'value1',",
    "  value2 = 'value2',",
    "};",
    "",
    "export default SomeEnum;",
    "",
  ],
  "example/models/SomeMaterializedView.ts": [
    "// @generated",
    "// This file is automatically generated by Kanel. Do not modify manually.",
    "",
    "/** Represents the materializedView test.some_materialized_view */",
    "export default interface SomeMaterializedView {",
    "  /** Database type: pg_catalog.int4 */",
    "  id: number | null;",
    "",
    "  /** Database type: pg_catalog.int4 */",
    "  some_other_table_id: number | null;",
    "}",
    "",
  ],
  "example/models/SomeOtherTable.ts": [
    "// @generated",
    "// This file is automatically generated by Kanel. Do not modify manually.",
    "",
    "/** Identifier type for some_other_table */",
    "export type Id = number & { __flavor?: 'Id' };",
    "",
    "/** Represents the table test.some_other_table */",
    "export default interface SomeOtherTable {",
    "  /** Database type: pg_catalog.int4 */",
    "  id: Id;",
    "}",
    "",
    "/** Represents the initializer for the table test.some_other_table */",
    "export interface SomeOtherTableInitializer {",
    "  /** Database type: pg_catalog.int4 */",
    "  id: Id;",
    "}",
    "",
    "/** Represents the mutator for the table test.some_other_table */",
    "export interface SomeOtherTableMutator {",
    "  /** Database type: pg_catalog.int4 */",
    "  id?: Id;",
    "}",
    "",
  ],
  "example/models/SomeRange.ts": [
    "// @generated",
    "// This file is automatically generated by Kanel. Do not modify manually.",
    "",
    "/** Represents the range test.some_range */",
    "type SomeRange = string;",
    "",
    "export default SomeRange;",
    "",
  ],
  "example/models/SomeTable.ts": [
    "// @generated",
    "// This file is automatically generated by Kanel. Do not modify manually.",
    "",
    "import type { Id } from './SomeOtherTable';",
    "",
    "/** Represents the table test.some_table */",
    "export default interface SomeTable {",
    "  /** Database type: pg_catalog.int4 */",
    "  id: number | null;",
    "",
    "  /** Database type: pg_catalog.int4 */",
    "  some_other_table_id: Id | null;",
    "}",
    "",
    "/** Represents the initializer for the table test.some_table */",
    "export interface SomeTableInitializer {",
    "  /** Database type: pg_catalog.int4 */",
    "  id?: number | null;",
    "",
    "  /** Database type: pg_catalog.int4 */",
    "  some_other_table_id?: Id | null;",
    "}",
    "",
    "/** Represents the mutator for the table test.some_table */",
    "export interface SomeTableMutator {",
    "  /** Database type: pg_catalog.int4 */",
    "  id?: number | null;",
    "",
    "  /** Database type: pg_catalog.int4 */",
    "  some_other_table_id?: Id | null;",
    "}",
    "",
  ],
  "example/models/SomeView.ts": [
    "// @generated",
    "// This file is automatically generated by Kanel. Do not modify manually.",
    "",
    "/** Represents the view test.some_view */",
    "export default interface SomeView {",
    "  /** Database type: pg_catalog.int4 */",
    "  id: number;",
    "",
    "  /** Database type: pg_catalog.int4 */",
    "  some_other_table_id: number;",
    "}",
    "",
  ],
  "example/models/some_function.ts": [
    "// @generated",
    "// This file is automatically generated by Kanel. Do not modify manually.",
    "",
    "export interface some_function_params {",
    "}",
    "",
  ],
  "example/models/some_procedure.ts": [
    "// @generated",
    "// This file is automatically generated by Kanel. Do not modify manually.",
    "",
    "export interface some_procedure_params {",
    "}",
    "",
  ],
}
`;
